src/Analysis.cpp

```diff
@@ -343,36 +382,83 @@ void doAliasReasoning(Instruction *instruction,
             return;
           }
 
-          ProgramVariable structVar = ProgramVariable(pointerOperand, index);
+          ProgramVariable structPV = ProgramVariable(pointerOperand);
+
+          ProgramVariable *originalStructPVRef =
+              programPoint->getPVRef(structPV.getCleanedName(), false);
+          ProgramVariable structVar =
+              ProgramVariable(originalStructPVRef->getValue(), index);
+
+          logout("spec index inst = " << *gepInst);
+          logout("specifying index for " << structVar.getCleanedName());
+
           programPoint->addAlias(sourceVar, structVar);
         }
       }
     }
   } else if (AllocaInst *allocate = dyn_cast<AllocaInst>(instruction)) {
-    if (llvm::StructType *structType =
-            llvm::dyn_cast<llvm::StructType>(allocate->getAllocatedType())) {
-      /*
-      I cannot find documentation that states how struct names appear in the IR,
-      but it seems that they typically look like "struct.[C struct name]"
-      TODO: prove/disprove that struct names are formatted as "struct.[C struct
-      name]"
-      */
-      std::string structName = structType->getName();
-      structName = sliceString(structName, structName.find_last_of('.') + 1,
-                               structName.size() - 1);
-      int numFields = structType->getNumElements();
-      for (int i = 0; i < numFields; i++) {
-
-        Annotation *anno = annotationHandler.getStructAnnotation(structName, i);
-        if (StructAnnotation *structAnno =
-                dynamic_cast<StructAnnotation *>(anno)) {
-          std::set<std::string> structMethods =
-              structAnno->getAnnotationMethods();
-          AnnotationType structAnnoType = structAnno->getAnnotationType();
-          ProgramVariable sourceVar = ProgramVariable(allocate, i);
-          programPoint->addVariable(sourceVar);
-        }
-      }
+    logout("alloca inst = " << *allocate);
+
+    StructType *structType =
+        llvm::dyn_cast<llvm::StructType>(allocate->getAllocatedType());
+
+    if (!structType && allocate->getAllocatedType()->isPointerTy()) {
+      structType = dataflow::unwrapValuePointerToStruct(allocate);
+    }
+
+    if (!structType) {
+      return;
+    }
+
+    programPoint->addVariable(ProgramVariable(allocate));
+
+    /*
+    I cannot find documentation that states how struct names appear in the IR,
+    but it seems that they typically look like "struct.[C struct name]"
+    TODO: prove/disprove that struct names are formatted as "struct.[C struct
+    name]"
```

Comment:
Turns out, there is no required format, but it mostly looks like "struct.[C struct name]." I'm not sure how else it could be formatted (I didn't see any other types of formatting in the docs), so to compromise, I'll use the debug information. There are DICompositeType tags in the debug information that specify a type and a name. The name tag in DICompositeType tags are more reliable and likely to align with the C code. So for each struct name `structName` we find, we'll search the debug info to see if there is a DICompositeType that specifies a struct with the name `structName`. If we don't find any, we'll error and adjust the code to handle it. 

I found one case where the struct formatting is not "struct.[C struct name]", and that was by manually editing the IR. Users will definitely not be editing the IR, so I won't worry about this case.

Reply:
This sounds like a good solution. Please make sure to update the comment in the code to explain this, too.